Created Issue Triage file (#166)
ansible-collections/dellemc.enterprise_sonic: docs/ISSUE_TRIAGE.md

Even if you don't have the time or the knowledge to investigate an issue we highly recommend that you upvote the issue if you happen to have the same problem. If you have further details that may help with investigating the issue please provide as much information as possible.

## External pull requests

Part of issue triage should also be triaging of external PRs. The main goal should be to make sure PRs from external contributors have an owner/reviewer and are not forgotten.

1. Check new external PRs which do not have a reviewer.
1. Check if there is a link to an existing issue.
1. If not and you know which issue it is solving, add the link yourself, otherwise ask the author to link the issue or create one.
1. Assign a reviewer based on who was handling the linked issue or what code or feature the PR touches (look at who was the last to make changes there if all else fails).

> ### GitHub Issue: Bug

Reviewer: Please delete this section. It seems redundant and doesn't fit well with the organization of the document.

The points described here can be incorporated into the "Bug Reports" section if desired.

Author: This section has been updated with flowchart as suggested by server storage team. Hence request to kindly review

Reviewer: I am marking this issue "Resolved" because the recent modifications make it no longer relevant.

## 3. Categorizing an issue

### Duplicate issues

Make sure it's not a duplicate by searching existing issues using related terms from the issue title and description. If you think the issue may be a potential duplicate and can't find the original existing issue, then please reach out to one of the maintainers and ask for help. If you identify that the issue is a duplicate of an existing issue:

1. Add a comment `duplicate of #<issue number>`
2. Add the `triage/duplicate` label

### Bug reports

If it's not perfectly clear that it's an actual bug, quickly try to reproduce it.

**It's a bug/it can be reproduced:**

1. Add a comment describing detailed steps for how to reproduce it, if applicable.
2. If maintainers aren't able to address the issue in a timely manner, then label the issue with `help wanted` and an optional `beginner friendly`. Also, include pointers to the problematic source code. Doing this allows us to solicit help from the community to get the issue resolved.
3. Move on to [prioritizing the issue](#4-prioritization-of-issues).

**It can't be reproduced:**

> 

Reviewer: This isn't grammatical and is a little hard to read. (Having two successive "Either" statements is confusing.)

Please delete the two "Either"s at the beginning of these items and do one of the following:

- Precede the two items with "Either", then add an "or" between the two items.
- Precede these two list items with "Do one of the following:" (This implies the Either/or.)

Author: Addressed as suggested

Reviewer: There are still two preceding "Either"s. Please delete the one for alternative "2".

Due to constraints, maintainers aren't always able to resolve issues in a timely manner, especially those of lower priority. For such issues, consider requesting help from the community. Use your best judgement. In general, requesting help from the community means that a contribution has a good chance of getting accepted and merged.

In many cases, the issue author or the community is most suitable to contribute changes since they're experts in their domain. Also, it's very common for someone to seek help from the community on a problem.

1. Kindly and politely add a comment which will notify the users subscribed to the issue of any updates.
   - Explain that the issue would be nice to get resolved, but it isn't prioritized to be worked on by maintainers for the foreseeable future.
   - If possible or applicable, try to help contributors get started by adding pointers to relevant source code or files as well as adding any helpful input on how to resolve the issue.
2. Label the issue with `help wanted`.
3. If applicable, label the issue with `beginner friendly` to denote that the issue is suitable for a beginner to work on.

## Investigation of issues

> When an issue has all basic information provided, but the reported problem cannot be reproduced at a first glance, the issue is labeled `triage/needs-information`. Depending on the perceived severity and/or number of upvotes, the investigation will either be delegated to another maintainer for further investigation or put on hold until someone else (maintainer or contributor) picks it up and eventually starts investigating it.

Reviewer: This wording is unclear. I will try to provide an alternate phrasing, but am somewhat guessing at what was intended. I think that if we ask people to label an issue "needs information" because it is difficult to reproduce, even though they do think it has enough information, we should explicitly ask them to explain why they are labeling it as "needs information". 

```
When an issue has all basic information provided, but the reported problem cannot be reproduced at a first glance, label the issue `triage/needs-information` and post a comment explaining why this label is being applied.
```

Author: Addressed as suggested

Even if you don't have the time or the knowledge to investigate an issue we highly recommend that you upvote the issue if you happen to have the same problem. If you have further details that may help with investigating the issue please provide as much information as possible.

## External pull requests

Part of issue triage should also be triaging of external PRs. The main goal should be to make sure PRs from external contributors have an owner/reviewer and are not forgotten.

1. Check new external PRs which do not have a reviewer.
1. Check if there is a link to an existing issue.
1. If not and you know which issue it is solving, add the link yourself, otherwise ask the author to link the issue or create one.

> 1. Assign a reviewer based on who was handling the linked issue or what code or feature does the PR touches (look at who was the last to make changes there if all else fails).

Reviewer: Delete the word "does".

> ## GitHub issue management workflow

Reviewer: Was there intended to be more content here? The "ask a question" and "new feature" types aren't shown, and no actual workflow steps are described.

Author: @kerry-meyer  :  As of now there is no additional content. If you have any suggestion for adding more content related to "Ask a Question" and/or "New Feature" types, kindly provide the content. 
Alternatively, you can approve this with the current content if that is acceptable.

Reviewer: Please at least provide the proposed additional steps of the workflow expected for handling of a "Bug" type issue (leading up to and including closing of the issue) and/or add a comment indicating that this is a placeholder for future content and stating that maintainers will provide additional details in the future. 

Alternatively, this section can be deleted since it is mostly already covered by the previous content of the document. Since the full workflow is beyond the scope of "triage", it could, and probably should be included in one of the other documents ("maintainer's guide" or "contributor's guide") if it's not already there and referenced here instead of providing a partial section.

Reviewer 2: I don't think we need this. Let's just delete it. 
If it's strictly required for OneAnsible, we can use the same content that other components have been using. Otherwise let's just get rid of it; it's extraneous detail.

Author: @jeff-yin This requirement is coming from OneAnsible. I am not clear on what you mean by content from "other components". Which other components are you referring to?

Author: @jeff-yin : This document is the guide for "core" maintainer/developer/lead who will do the first round of triaging. Do you mean to remove this document completely? Or, do you mean to remove this particular point?

Reviewer 2: "other components" -- just meaning Storage and Servers.

I just mean to remove this particular point, not the entire section. I wrote this in reply to this thread about the "GitHub issue management workflow" section, so I'm not sure why it was interpreted that I meant to remove the document completely.

Author: Server/storage team is planning to create a workflow diagram and add to this document later. Until then, the section "##GitHub issue management workflow" is not required and hence removed. This document will be enhanced in future version when the diagram is available.

Reviewer: Because real content has been provided, I am marking this issue "Resolved".

<!--
Copyright (c) 2022 Dell Inc., or its subsidiaries. All Rights Reserved.

Licensed under the GPL, Version 3.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://www.gnu.org/licenses/gpl-3.0.txt
-->

# Triage issues

The main goals of an issue triage are to categorize open issues based on priority and to ensure that these issues contain adequate information for contributors to work on them.

> **Note:** This information is for project Maintainers, Owners, and Admins. If you are a Contributor, then you will not be able to perform most of the tasks in this topic.

The core maintainers of this project are responsible for categorizing all incoming issues and delegating any critical or important issue to other maintainers. Triage provides an important way to contribute to an open source project.

A triage helps resolve issues quickly by :

- Ensuring the intent of an issue is clear. The issue should clearly describe the problem the end user is experiencing as well as the steps to reproduce the problem.
- Giving a contributor the information they need before they commit to resolving an issue.

> - Lowering the issue count by preventing duplicate issues.

Reviewer 3: ```suggestion

```

Author: removed as suggested

## 1. Find issues that need triage

The easiest way to find issues that haven't been triaged is to search for issues with the `needs-triage` label.

## 2. Ensure the issue contains basic information

When creating a new issue, the author must provide all of the information required by the project's GitHub issue template

### Standard issue information that must be included

The following section describes the various issue templates and the expected content.

#### Bug reports

A bug report should contain explanation of what the problem is, what the expected outcome is, and how to reproduce the problem. Additionally, any other supporting information such as screenshots, console logs, and environment details should be included in the bug report. 

 - Product Name/Version
 - Component/Module Name
 - Ansible Collection Version	
 - Configuration
 - Steps to Reproduce
 - Expected/Actual Behavior
 - Logs/Screenshots
 - Any other additional information...

#### Feature requests

Should explain what feature that the author wants to be added and why that is needed.

#### Ask a Question requests

> In general, if the issue description and title is perceived as a question no more information is needed.

Author: Addressed as suggested

> In general, if the issue description and title is perceived as a question no more information is needed.

Reviewer: Thanks for correcting this.

Resolved.

2. Either [delegate further investigations](#investigation-of-issues) to someone else.  Provide details in a comment.

**It works as intended/by design:**

1. Kindly and politely add a comment explaining briefly why we think it works as intended and close the issue.
2. Label the issue `triage/works-as-intended`.
3. Remove the `needs-triage` label.

**It does not work as intended/by design:**

### Feature requests

1. If the feature request does not align with the product vision, add a comment indicating so, remove the `needs-triage` label and close the issue
2. Otherwise, move on to [prioritizing the issue](#4-prioritization-of-issues).  Assign the appropriate priority label to the issue, add the appropriate comments to the issue, and remove the `needs-triage` label.

## 4. Prioritization of issues
In general, bugs and feature request issues should be labeled with a priority.

This can be the most difficult task when triaging issues since it requires a lot of knowledge, context, and experience before being able to start feeling comfortable adding a certain priority label.

> In order to gain comfort with prioritizing issues, consulting with experienced project members on issues is highly encouraged. By doing that you learn more and eventually be more and more comfortable with prioritizing issues.

Reviewer 3: ```suggestion
In order to gain comfort with prioritizing issues, consulting with experienced project members on issues is highly encouraged.
```

Author: Addressed as suggested

2. Label the issue `priority/critical`.
3. Escalate the problem to the maintainers.
4. Assign the issue or ask a maintainer for help assigning the issue to someone as their immediate top priority.
5. Add the issue to the next upcoming release milestone.

### High priority

1. Label the issue `priority/high`.
2. Add the issue to the next upcoming release milestone.
3. Prioritize it or assign someone to work on it now or very soon.
4. Consider requesting [help from the community](#5-requesting-help-from-the-community).

### Low priority

1. If the issue is deemed possibly useful but is a low priority, then label the issue `priority/low`.
2. The functional impact of an issue determines how high of a priority the issue is.
3. Consider requesting [help from the community](#5-requesting-help-from-the-community).

## 5. Requesting help from the community

Due to constraints, maintainers aren't always able to resolve issues in a timely manner, especially those of lower priority. For such issues, consider requesting help from the community. Use your best judgement. In general, requesting help from the community means that a contribution has a good chance of getting accepted and merged.

> In many cases, the issue author or the community is most suitable to contribute changes since they're experts in their domain. It's also quite common that someone has tried to get something to work using the documentation without success and made an effort to get it to work and/or reached out to the community to get the missing information.

Reviewer 3: ```suggestion
In many cases, the issue author or the community is most suitable to contribute changes since they're experts in their domain. Also, it's very common for someone to seek help from the community on a problem.
```

Author: Addressed as suggested

 - Product Name/Version
 - Component/Module Name
 - Ansible Collection Version	
 - Configuration
 - Steps to Reproduce
 - Expected/Actual Behavior
 - Logs/Screenshots
 - Any other additional information...

#### Feature requests

Should explain what feature that the author wants to be added and why that is needed.

#### Ask a Question requests

In general, if the issue description and title are perceived as questions no more information is needed.

### Good practices

To make issue handling easier for everyone, it is suggested to:

- Make sure that issue titles are named to explain the subject of the issue, have correct spelling, and don't include irrelevant information and/or sensitive information.
- Make sure that issue descriptions don't include irrelevant information.
- Make sure that issues do not contain sensitive information.
- Make sure that issues have all relevant fields filled in.
- If an issue is unclear, then try to edit the title and description for more clarity or leave a comment requesting that edits to the issue be made.

> > **Note:** Above rules are applicable to both new and existing issues.

Reviewer 2: I don't think this line is needed.

Author: Okay. We will discuss with server storage team and will get back to you.

Reviewer 2: I just think that saying "rules are applicable to both new and existing issues" is already implied (why would they not be applicable?) and doesn't really add any value to the document.

Author: Removed as suggested.

2. Add the `triage/duplicate` label

### Bug reports

If it's not perfectly clear that it's an actual bug, quickly try to reproduce it.

**It's a bug/it can be reproduced:**

1. Add a comment describing detailed steps for how to reproduce it, if applicable.
2. If maintainers aren't able to address the issue in a timely manner, then label the issue with `help wanted` and an optional `beginner friendly`. Also, include pointers to the problematic source code. Doing this allows us to solicit help from the community to get the issue resolved.
3. Move on to [prioritizing the issue](#4-prioritization-of-issues).

**It can't be reproduced:**

1. Either request that [more information](#2-ensure-the-issue-contains-basic-information) is needed to investigate the issue more thoroughly. Provide details in a comment.
or
2. Either [delegate further investigations](#investigation-of-issues) to someone else.  Provide details in a comment.

**It works as intended/by design:**

1. Kindly and politely add a comment explaining briefly why we think it works as intended and close the issue.
2. Label the issue `triage/works-as-intended`.
3. Remove the `needs-triage` label.

> **It does not work as intended/by design:**

Reviewer: Please add some description here or at least add a comment that indicates an action item for maintainers to provide the additional information. There is already a suitable matching action described for feature requests (prioritization and removal of the 'needs-triage' label). That could be copied here to complete this section.

Author: We will discuss with server storage team and get back to you.

Reviewer 2: If something is both "not a bug" but also "does not work as intended/by design" then it seems appropriate to convert/tag the issue into a Feature Request. Perhaps you can add language suggesting this.

Author: @jeff-yin we will discuss with server storage team and will get back to you.

Author: Yes, this comment was discussed with server/storage team. It is agreed that "It does not work as intended/by design" is same as "It's a bug/it can be reproduced". Hence the additional category/line "It does not work as intended/by design" has been removed.

Even if you don't have the time or the knowledge to investigate an issue we highly recommend that you upvote the issue if you happen to have the same problem. If you have further details that may help with investigating the issue please provide as much information as possible.

## External pull requests

Part of issue triage should also be triaging of external PRs. The main goal should be to make sure PRs from external contributors have an owner/reviewer and are not forgotten.

1. Check new external PRs which do not have a reviewer.
1. Check if there is a link to an existing issue.
1. If not and you know which issue it is solving, add the link yourself, otherwise ask the author to link the issue or create one.
1. Assign a reviewer based on who was handling the linked issue or what code or feature the PR touches (look at who was the last to make changes there if all else fails).

## GitHub issue management workflow

> The following section describes the triage workflow for new GitGHub issues that get created.

Reviewer 2: "Github" is misspelled here.

Author: This section has been removed as per above discussion thread.

## External pull requests

Part of issue triage should also be triaging of external PRs. The main goal should be to make sure PRs from external contributors have an owner/reviewer and are not forgotten.

1. Check new external PRs which do not have a reviewer.
1. Check if there is a link to an existing issue.
1. If not and you know which issue it is solving, add the link yourself, otherwise ask the author to link the issue or create one.
1. Assign a reviewer based on who was handling the linked issue or what code or feature the PR touches (look at who was the last to make changes there if all else fails).

## GitHub issue management workflow

The following section describes the triage workflow for new GitGHub issues that get created.

### GitHub Issue: Bug

This workflow starts off with a GitHub issue of type bug being created.

1. Collaborator or maintainer creates a GitHub bug using the appropriate GitHub issue template
2. By default a bug will be created with the `type/bug` and `needs-triage` labels

The following flowchart defines the workflow,

> ```                                                                                                                                                     

Reviewer 2: This should just be a PNG that is attached in the doc. Not sure why it needs to be ASCII art. But it's ok for now.

Reviewer: I'm resolving this conversation because Jeff has approved the PR.

   +----------------------------------+  NO   +--------------|-------------+                                                                             
   | label: triage/needs-information  ---------  All required information  |                                                                             
   |                                  |       |  contained in issue?       |                                                                             
   +-----------------------------|----+       +--------------|-------------+                                                                             
                                 |                           | YES                                                                                       
                                 |                           |                                                                                           
   +--------------------------+  |                +---------------------+ YES +---------------------------------------+                                  

>    |label:                    |  |                |  Dupicate Issue?    ------- Comment `Duplicate of #<issue number>`                                   

Reviewer: typo/spelling: Change "Dupicate" to "Duplicate".

Author: addressed

                 |                             |NO                                          |                                                            
                 |                             |                         +------------------|------------------+                                         
    +------------|-------------+          +----|----------------+ YES    |  Add details to issue               |                                         
    |                          ------------  Signal Community?  ----------  label: help wanted                 |                                         
    |Remove needs-triage label |          |                     |        |  label: beginner friendly (optional)|                                         
    +--------------------------+          +---------------------+        +-------------------------------------+                                          
                                                                                                                                                                                       

> ``` 

Reviewer: Shouldn't there be a workflow title line here, similar to the one for "Bug" type issues?

Author: Updated as suggested

        +-|---------|---+       +--------+    | Remove needs-triage label    |                                                   
        |  Close issue  |       |  Done  ------ Remove type/feature-request  |                                                   
        |               |       |        |    | milestone?                   |                                                   
        +---------------+       +--------+    +------------------------------+           
```
If the author does not respond to a request for more information within the timespan of a week, close the issue with a kind note stating that the author can request for the issue to be reopened when the necessary information is provided.

In some cases you may receive a request you do not wish to accept.  Perhaps the request doesn't align with the project scope or vision.  It is important to tactfully handle contributions that don't meet the project standards.

1. Acknowledge the person behind the contribution and thank them for their interest and contribution
2. Explain why it didn't fit into the scope of the project or vision

> 3. Don't leave an unwanted contributions open.  Immediately close the contribution you do not wish to accept

Reviewer: Change "contribution" to "request" in "Immediately close the contribution ..."

Author: Updated as suggested

Reviewer: There was another instance of the same thing in the referenced line. Please make the same change for that instance and other similar instances in other lines. See the new comment for this below specifying the current line numbers that need this change.

                 |                             |NO                                          |                                                            
                 |                             |                         +------------------|------------------+                                         
    +------------|-------------+          +----|----------------+ YES    |  Add details to issue               |                                         
    |                          ------------  Signal Community?  ----------  label: help wanted                 |                                         
    |Remove needs-triage label |          |                     |        |  label: beginner friendly (optional)|                                         
    +--------------------------+          +---------------------+        +-------------------------------------+                                          
                                                                                                                                                                                       

> ```

Author: We are able to see the headings separately. The content for the first flowchart starts in line : 196 followed by document marker at line : 200 to mark the start of first flowchart. Then the document marker ends at line : 246 for the completion of first flow chart. After which the heading for second flow chart is given at line : 248, then the document marker starts at line : 250 to mark the start of second flowchart. The document marker ends at line : 281 for the completion of second flow chart. Would request to verify once in this [md file format](https://github.com/ansible-collections/dellemc.enterprise_sonic/blob/ed462e20322506c2050f6042718dbbc7b4615d7a/docs/ISSUE_TRIAGE.md).

> ```

Reviewer: Okay. I don't know why I thought the second heading wasn't being displayed. I see that it is.

I will resolve this last issue, approve the PR, and merge it.

          |         |                                         |                                                                  
          |         |                         +---------------|--------------+                                                                                                    
          |         |                         | label: type/feature          |                                                   
        +-|---------|---+       +--------+    | Remove needs-triage label    |                                                   
        |  Close issue  |       |  Done  ------ Remove type/feature-request  |                                                   
        |               |       |        |    | milestone?                   |                                                   
        +---------------+       +--------+    +------------------------------+           
```
If the author does not respond to a request for more information within the timespan of a week, close the issue with a kind note stating that the author can request for the issue to be reopened when the necessary information is provided.

> In some cases you may receive a request you do not wish to accept.  Perhaps the request doesn't align with the project scope or vision.  It is important to tactfully handle contributions that don't meet the project standards.

Reviewer: Please also fix the other instances where the word "contributions"  appears instead of "request".  There's one in this line and also instances on lines 286 and 288.
